File: my-folder/0211-design-add-and-search-words-data-structure/solution.py
class TrieNode():
    def __init__(self):
        self.children={}
        self.word=False
class WordDictionary:
    def __init__(self):
        self.root=TrieNode()
    def addWord(self, word: str) -> None:
        curr=self.root
        for w in word:
            if w not in curr.children:
                curr.children[w]=TrieNode()
            curr=curr.children[w]
        curr.word=True    

    def search(self, word: str) -> bool:
        def recsearch(i,root):
            curr=root
            for j in range(i,len(word)):
                c=word[j]
                if c=='.':
                    for child in curr.children.values():
                        if recsearch(j+1,child):
                            return True
                    return False
                else:
                    if c not in curr.children:
                        return False
                    curr=curr.children[c] 
            return curr.word
        return recsearch(0,self.root)

File: my-folder/0211-design-add-and-search-words-data-structure/solution.java
class TrieNode{
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEnd = false;
}
class WordDictionary {
    TrieNode root;
    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode curr = root;
        for(char c : word.toCharArray()){
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        curr.isEnd = true;
    }
    
    public boolean search(String word) {
        return dfs(0, word, root);
    }

    private boolean dfs(int idx, String word, TrieNode root){
        TrieNode node = root;

        for(int i = idx; i < word.length(); i++){
            char c = word.charAt(i);
            if(c == '.'){
                for(TrieNode child : node.children.values()){
                    if(child != null && dfs(i + 1, word, child)){
                        return true;
                    }
                }
                return false;
            }
            else{
                if(!node.children.containsKey(c)){
                    return false;
                }

                node = node.children.get(c);
            }
        }

        return node.isEnd;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
